File: backend/app/open_node/domain/node_topologies.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator


class NodeTopologyPoint(BaseModel):
    model_config = {"extra": "forbid"}

    x: float = Field(ge=-100_000, le=100_000, allow_inf_nan=False)
    y: float = Field(ge=-100_000, le=100_000, allow_inf_nan=False)


class NodeTopologyStage(BaseModel):
    model_config = {"extra": "forbid"}

    node_ids: list[UUID] = Field(min_length=1, max_length=16)
    load_balance_strategy: Literal["round-robin"] = "round-robin"

    @field_validator("node_ids")
    @classmethod
    def unique_nodes(cls, value: list[UUID]) -> list[UUID]:
        if len(set(value)) != len(value):
            raise ValueError("A topology stage cannot contain duplicate nodes")
        return value


class NodeTopologyWrite(BaseModel):
    model_config = {"extra": "forbid"}

    name: str = Field(min_length=1, max_length=120)
    enabled: bool = Field(default=True, strict=True)
    stages: list[NodeTopologyStage] = Field(min_length=2, max_length=8)
    layout: dict[str, NodeTopologyPoint] = Field(default_factory=dict, max_length=128)

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized or any(ord(char) < 32 for char in normalized):
            raise ValueError("Topology name is invalid")
        return normalized

    @model_validator(mode="after")
    def validate_shape(self):
        flattened = [node_id for stage in self.stages for node_id in stage.node_ids]
        if len(set(flattened)) != len(flattened):
            raise ValueError("A node can appear only once in a topology")
        if len(self.stages[-1].node_ids) != 1:
            raise ValueError("The final topology stage must contain exactly one exit node")
        known = {str(node_id) for node_id in flattened}
        if set(self.layout) - known:
            raise ValueError("Topology layout references an unknown node")
        return self


class NodeTopologyCreate(NodeTopologyWrite):
    pass


class NodeTopologyUpdate(NodeTopologyWrite):
    expected_revision: str = Field(pattern=r"^[a-f0-9]{64}$")


class NodeTopologyDelete(BaseModel):
    model_config = {"extra": "forbid"}

    expected_revision: str = Field(pattern=r"^[a-f0-9]{64}$")
    confirm_name: str = Field(min_length=1, max_length=120)


class NodeTopologyCandidate(BaseModel):
    id: UUID
    name: str
    server_id: UUID
    server_name: str
    server_kind: Literal["direct", "leased-line", "residential"]
    protocol: str


class NodeTopologyRead(NodeTopologyWrite):
    id: UUID
    revision: str = Field(pattern=r"^[a-f0-9]{64}$")
    created_at: datetime
    updated_at: datetime


class NodeTopologiesResponse(BaseModel):
    topologies: list[NodeTopologyRead]
    candidates: list[NodeTopologyCandidate]
    license_required: Literal[False] = False


class NodeTopologyResponse(BaseModel):
    topology: NodeTopologyRead
    license_required: Literal[False] = False

File: frontend/src/domain/node-topologies.ts
export interface NodeTopologyCandidate {
  id: string;
  name: string;
  server_id: string;
  server_name: string;
  server_kind: "direct" | "leased-line" | "residential";
  protocol: string;
}

export interface NodeTopologyPoint {
  x: number;
  y: number;
}

export interface NodeTopologyStage {
  node_ids: string[];
  load_balance_strategy: "round-robin";
}

export type NodeTopologyLayout = Record<string, NodeTopologyPoint>;

export interface NodeTopology {
  id: string;
  name: string;
  enabled: boolean;
  stages: NodeTopologyStage[];
  layout: NodeTopologyLayout;
  revision: string;
  created_at: string;
  updated_at: string;
}

export interface NodeTopologiesResponse {
  topologies: NodeTopology[];
  candidates: NodeTopologyCandidate[];
  license_required: false;
}

export interface NodeTopologyWrite {
  name: string;
  enabled: boolean;
  stages: NodeTopologyStage[];
  layout: NodeTopologyLayout;
}

export interface NodeTopologyUpdate extends NodeTopologyWrite {
  expected_revision: string;
}

export interface NodeTopologyDelete {
  expected_revision: string;
  confirm_name: string;
}

export interface TopologyDraftValidation {
  valid: boolean;
  errors: string[];
}

export interface TopologyDraftMutation {
  stages: NodeTopologyStage[];
  error: string | null;
}

export function copyTopologyStages(stages: NodeTopologyStage[]): NodeTopologyStage[] {
  return stages.map(stage => ({
    node_ids: [...stage.node_ids],
    load_balance_strategy: "round-robin",
  }));
}

function nodes(stages: NodeTopologyStage[]) {
  return stages.flatMap(stage => stage.node_ids);
}

export function validateTopologyDraft(
  name: string,
  stages: NodeTopologyStage[],
  candidates: NodeTopologyCandidate[],
): TopologyDraftValidation {
  const errors: string[] = [];
  const normalizedName = name.trim();
  if (!normalizedName) errors.push("请输入编排名称。");
  else if (normalizedName.length > 120) errors.push("编排名称不能超过 120 个字符。");
  if (stages.length < 2) errors.push("节点编排至少需要 2 跳。");
  if (stages.length > 8) errors.push("节点编排最多支持 8 跳。");
  stages.forEach((stage, index) => {
    if (!stage.node_ids.length) errors.push(`第 ${index + 1} 跳至少需要一个节点。`);
    if (stage.node_ids.length > 16) errors.push(`第 ${index + 1} 跳最多支持 16 个节点。`);
    if (stage.load_balance_strategy !== "round-robin") errors.push(`第 ${index + 1} 跳仅支持轮询负载均衡。`);
  });
  if (stages.length && stages.at(-1)!.node_ids.length !== 1) {
    errors.push("最终出口必须且只能包含一个节点。");
  }

  const byId = new Map(candidates.map(candidate => [candidate.id, candidate]));
  const nodeIds = nodes(stages);
  const seenNodes = new Set<string>();
  const seenServers = new Set<string>();
  for (const nodeId of nodeIds) {
    if (seenNodes.has(nodeId)) {
      errors.push("同一个节点不能在编排中重复出现。");
      break;
    }
    seenNodes.add(nodeId);
  }
  for (const nodeId of nodeIds) {
    const candidate = byId.get(nodeId);
    if (!candidate) {
      errors.push("编排包含已停用或不存在的节点，请移除后再保存。");
      continue;
    }
    if (seenServers.has(candidate.server_id)) {
      errors.push("同一台服务器不能重复经过，已阻止回还环路。");
      break;
    }
    seenServers.add(candidate.server_id);
  }
  return { valid: errors.length === 0, errors: [...new Set(errors)] };
}

export function insertTopologyCandidate(
  stages: NodeTopologyStage[],
  candidate: NodeTopologyCandidate,
  candidates: NodeTopologyCandidate[],
  stageIndex: number | null,
): TopologyDraftMutation {
  const used = new Set(nodes(stages));
  if (used.has(candidate.id)) return { stages, error: `节点“${candidate.name}”已经在编排中。` };
  const byId = new Map(candidates.map(item => [item.id, item]));
  const reusedServer = [...used].map(id => byId.get(id)).find(item => item?.server_id === candidate.server_id);
  if (reusedServer) {
    return { stages, error: `服务器“${candidate.server_name}”已经经过，不能形成重复路径或回还环路。` };
  }
  if (stageIndex === null) {
    if (stages.length >= 8) return { stages, error: "节点编排最多支持 8 跳。" };
    return { stages: [...copyTopologyStages(stages), { node_ids: [candidate.id], load_balance_strategy: "round-robin" }], error: null };
  }
  if (!Number.isInteger(stageIndex) || stageIndex < 0 || stageIndex >= stages.length) {
    return { stages, error: "目标跳数已经变化，请重新选择。" };
  }
  if (stages[stageIndex].node_ids.length >= 16) return { stages, error: "每一跳最多支持 16 个节点。" };
  const updated = copyTopologyStages(stages);
  updated[stageIndex].node_ids.push(candidate.id);
  return { stages: updated, error: null };
}

export function removeTopologyNode(stages: NodeTopologyStage[], nodeId: string): NodeTopologyStage[] {
  return stages
    .map(stage => ({ ...stage, node_ids: stage.node_ids.filter(id => id !== nodeId) }))
    .filter(stage => stage.node_ids.length > 0);
}

export function removeTopologyStage(stages: NodeTopologyStage[], index: number): NodeTopologyStage[] {
  return copyTopologyStages(stages).filter((_stage, stageIndex) => stageIndex !== index);
}

export function reorderTopologyStage(stages: NodeTopologyStage[], from: number, to: number): NodeTopologyStage[] {
  if (!Number.isInteger(from) || !Number.isInteger(to) || from < 0 || to < 0
    || from >= stages.length || to >= stages.length || from === to) return copyTopologyStages(stages);
  const updated = copyTopologyStages(stages);
  const [moved] = updated.splice(from, 1);
  updated.splice(to, 0, moved);
  return updated;
}

export function topologyLayoutForStages(layout: NodeTopologyLayout, stages: NodeTopologyStage[]) {
  const selected = new Set(nodes(stages));
  return Object.fromEntries(Object.entries(layout).filter(([nodeId]) => selected.has(nodeId)));
}
